keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -153,6 +153,18 @@ Additionally, Kubelet is authorized to update PodSpec, and NodeRestriction
 admission plugin is extended to limit Kubelet's update access only to Pod's
 ResourcesAllocated field for CPU and memory resources.
 
+<<[UNRESOLVED]>>
+Why is "allocated" spec?  The need for special admission control is a strong
+smell.  Why doesn't this belong in status?  If there's a compelling reason,
+this KEP should say so.
+<<[/UNRESOLVED]>>
+
+<<[UNRESOLVED]>>
+Do we want to define a subresource for setting `resources` on pods?  This would
+allow us to grant RBAC access to things like VPA without allowing full write
+access to pod spec.
```

Comment:
Adding subresource was our initial proposal, based on the /binding precedent. At the [Live API review session](https://youtu.be/faRARV3C7Fk?t=2974) in K8s contributor summit in San Diego, @liggitt said adding ResourcesAllocated was reasonable, but in the post-session chat, we all felt that we were starting to add subresources willy-nilly.

And after playing with admission controller, it offers good advantages - for e.g ability to handle down-level clients who may drop ResourcesAllocated on write in addition to access control.

Reply:
Sorry, we're talking about 2 things.  Ignore "allocated" for this thread. :)

I'm asking about a sub-resource for `pod.spec.containers[*].resources` which would allow things like VPA to be granted access to set resources but not anything else in `pod.spec`.

Reply:
VPA is like any other external user in this capacity. And companies may have their own solutions in-lieu of VPA. In any case, how does it help us if we don't have ResourcesAllocated or another similar alternative?

Reply:
Give VPA-ish things PUT/PATCH access to `pods/resources` and GET/LIST/WATCH access to `pods`.

They can read the status for "actual" resources and write the /resources subresource to request a resize.

Without the sub-resource you have to grant access to VPA-ish things to PUT/PATCH all of `pod` spec - that seems worse in every way?

Reply:
We are limiting what can be changed by the user in validation. There are a limited set of fields that users can patch. e.g. image, annotations. There's no subresource for changing those fields.

Reply:
You're missing my point.  Let's assume that Pod spec is immutable except for fields `X`, `Y`, `Z`, and `resources`.

Users own their own pods.  They are allowed to change any of those 4 fields (and any others that become mutable over time).  This is granted by "update" and "patch" verbs on the "pods" resource.

VPA and arbitrary VPA-ish things need to change `resources`.  Granting VPA access to change `X`, `Y`, or `Z` is a Bad Thing.  So we should consider a subresource that ONLY grants access to `resources`.

To restrict users' "update" rights to a subset of mutable fields is, at best, weird.  To do that outside the scope of RBAC is just wrong, IMO.

Reply:
Fair. But how would we define VPA and friends in RBAC? (Sorry I've only touched RBAC very superficially) It seems they would look like any other user, I suspect it will functionally end up no different than PATCH.

Reply:
VPA has a service-account which is linked to a role-binding which ties it to a role, which has RBAC grants.  Every component should have a unique service-account and the smallest set of permissions possible.

Take, for example system:kube-scheduler:

```
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - delete
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - bindings
  - pods/binding
  verbs:
  - create
```

Notice that the scheduler can create a binding, but can not patch or update the rest of the pod.

Reply:
I need to wrap my head around RBAC from entities external to K8s, but if I'm on the right track, it would be something like https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/deploy/vpa-rbac.yaml ? 

Any Vertical_Pod_Scaler_XYZ entity has a way to access say ```podresources``` subresource allowing it to set resources without needing any changes in https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go ?

Reply:
I think that's right - there's lots of ways that RBAC can get configured, but the main point is that VPA doesn't get FULL access to write to pod.spec

Reply:
Also, after speaking with @liggitt we realized that this is not a pod-specific thing.  If we defined a `/resize` subresource, that could apply to Deployments, ReplicaSets, etc too.

Reply:
Are there any objections to defining a subresource for this?

Reply:
> Are there any objections to defining a subresource for this?

Not from me. I'm all for it. In our original user scenario a Job spawns the Pods, and we could benefit from ability to resize Job spec based on external triggers.